File: stereo3_SGM.m
% Stereo Matching using Semi-Global Matching
% Computes a disparity map from a rectified stereo pair using Semi-Global Matching

% Set parameters
dispLevels = 16; %disparity range: 0 to dispLevels-1
lambda = 5; %weight of smoothness cost
trunc = 4; %truncation of smoothness cost

% Define matching cost function
computeMatchingCost = @(left,right) abs(left-right); %absolute differences

% Define smoothness cost function
computeSmoothnessCost = @(d1,d2) lambda*min(abs(d1-d2),trunc);

% Load left and right images in grayscale
leftImg = rgb2gray(imread('left.png'));
rightImg = rgb2gray(imread('right.png'));

% Apply a Gaussian filter
leftImg = imgaussfilt(leftImg,0.6,'FilterSize',5);
rightImg = imgaussfilt(rightImg,0.6,'FilterSize',5);

% Get the size
[rows,cols] = size(leftImg);

% Convert to int32
leftImg = int32(leftImg);
rightImg = int32(rightImg);

% Compute pixel-based matching costs
matchingCosts = zeros(rows,cols,dispLevels,'int32');
for d = 0:dispLevels-1
    rightImgShifted = circshift(rightImg,d,2);
    matchingCosts(:,:,d+1) = computeMatchingCost(leftImg,rightImgShifted);
end

% Compute smoothness costs
d = 0:dispLevels-1;
smoothnessCosts = computeSmoothnessCost(d,d.');
smoothnessCosts3H = zeros(1,dispLevels,dispLevels,'int32');
smoothnessCosts3H(1,:,:) = smoothnessCosts;
smoothnessCosts3V = permute(smoothnessCosts3H,[2 1 3]);

% Initialize minimum cost paths for the 4 directions
fromLeft = zeros(rows,cols,dispLevels,'int32');
fromRight = zeros(rows,cols,dispLevels,'int32');
fromUp = zeros(rows,cols,dispLevels,'int32');
fromDown = zeros(rows,cols,dispLevels,'int32');

% Compute minimum cost paths for left to right direction
for x = 1:cols-1
    sumCosts = matchingCosts(:,x,:) + fromLeft(:,x,:) + smoothnessCosts3H;
    minSumCosts = min(sumCosts,[],3);
    normalizedCosts = minSumCosts - min(minSumCosts,[],2);
    fromLeft(:,x+1,:) = normalizedCosts;
end

% Compute minimum cost paths for right to left direction
for x = cols:-1:2
    sumCosts = matchingCosts(:,x,:) + fromRight(:,x,:) + smoothnessCosts3H;
    minSumCosts = min(sumCosts,[],3);
    normalizedCosts = minSumCosts - min(minSumCosts,[],2);
    fromRight(:,x-1,:) = normalizedCosts;
end

% Compute minimum cost paths for up to down direction
for y = 1:rows-1
    sumCosts = matchingCosts(y,:,:) + fromUp(y,:,:) + smoothnessCosts3V;
    minSumCosts = min(sumCosts,[],3).';
    normalizedCosts = minSumCosts - min(minSumCosts,[],2);
    fromUp(y+1,:,:) = normalizedCosts;
end

% Compute minimum cost paths for down to up direction
for y = rows:-1:2
    sumCosts = matchingCosts(y,:,:) + fromDown(y,:,:) + smoothnessCosts3V;
    minSumCosts = min(sumCosts,[],3).';
    normalizedCosts = minSumCosts - min(minSumCosts,[],2);
    fromDown(y-1,:,:) = normalizedCosts;
end

% Compute total costs
totalCosts = fromLeft + fromRight + fromUp + fromDown;

% Compute the disparity map
[~,ind] = min(totalCosts,[],3);
dispMap = ind-1;

% Normalize the disparity map for display
scaleFactor = 256/dispLevels;
dispImg = uint8(dispMap*scaleFactor);

% Show disparity map
figure; imshow(dispImg)

% Save disparity map
imwrite(dispImg,'disparity3_SGM.png')

File: python/stereo3_SGM.py
# Stereo Matching using Semi-Global Matching
# Computes a disparity map from a rectified stereo pair using Semi-Global Matching

import numpy as np
import cv2 as cv
import matplotlib.pyplot as plt

# Set parameters
dispLevels = 16 #disparity range: 0 to dispLevels-1
lambda_ = 5 #weight of smoothness cost
trunc = 4 #truncation of smoothness cost

# Define matching cost function
computeMatchingCost = lambda left,right: np.absolute(left-right) #absolute differences

# Define smoothness cost function
computeSmoothnessCost = lambda d1,d2: lambda_*np.minimum(np.absolute(d1-d2),trunc)

# Compute minimum cost paths
def computeMinSumCosts(costs):
    sumCosts = costs[:,:,:,np.newaxis] + smoothnessCosts4d
    minSumCosts = np.amin(sumCosts,axis=2)
    minSumCosts = minSumCosts - np.amin(minSumCosts,axis=2)[:,:,np.newaxis] #normalize costs
    return minSumCosts

# Load left and right images in grayscale
leftImg = cv.imread("left.png",cv.IMREAD_GRAYSCALE)
rightImg = cv.imread("right.png",cv.IMREAD_GRAYSCALE)

# Apply a Gaussian filter
leftImg = cv.GaussianBlur(leftImg,(5,5),0.6)
rightImg = cv.GaussianBlur(rightImg,(5,5),0.6)

# Get the size
(rows,cols) = leftImg.shape

# Convert to int32
leftImg = leftImg.astype(np.int32)
rightImg = rightImg.astype(np.int32)

# Compute pixel-based matching costs
matchingCosts = np.zeros((rows,cols,dispLevels),dtype=np.int32)
for d in range(dispLevels):
    rightImgShifted = np.roll(rightImg,d,1)
    matchingCosts[:,:,d] = computeMatchingCost(leftImg,rightImgShifted)

# Compute smoothness costs
d = np.arange(dispLevels)
smoothnessCosts = computeSmoothnessCost(d,d[np.newaxis,:].T)
smoothnessCosts4d = smoothnessCosts[np.newaxis,np.newaxis,:,:].astype(np.int32)

# Initialize minimum cost paths for the 4 directions
fromLeft = np.zeros((rows,cols,dispLevels),dtype=np.int32)
fromRight = np.zeros((rows,cols,dispLevels),dtype=np.int32)
fromUp = np.zeros((rows,cols,dispLevels),dtype=np.int32)
fromDown = np.zeros((rows,cols,dispLevels),dtype=np.int32)

# Compute minimum cost paths for left to right direction
for x in range(cols-1):
    costs = (matchingCosts[:,x,:] + fromLeft[:,x,:])[:,np.newaxis,:]
    fromLeft[:,x+1,:] = computeMinSumCosts(costs)[:,0,:]

# Compute minimum cost paths for right to left direction
for x in range(cols-1,0,-1):
    costs = (matchingCosts[:,x,:] + fromRight[:,x,:])[:,np.newaxis,:]
    fromRight[:,x-1,:] = computeMinSumCosts(costs)[:,0,:]

# Compute minimum cost paths for up to down direction
for y in range(rows-1):
    costs = (matchingCosts[y,:,:] + fromUp[y,:,:])[np.newaxis,:,:]
    fromUp[y+1,:,:] = computeMinSumCosts(costs)[0,:,:]

# Compute minimum cost paths for down to up direction
for y in range(rows-1,0,-1):
    costs = (matchingCosts[y,:,:] + fromDown[y,:,:])[np.newaxis,:,:]
    fromDown[y-1,:,:] = computeMinSumCosts(costs)[0,:,:]

# Compute total costs
totalCosts = fromLeft + fromRight + fromUp + fromDown

# Compute the disparity map
dispMap = np.argmin(totalCosts,axis=2)

# Normalize the disparity map for display
scaleFactor = 256/dispLevels
dispImg = (dispMap*scaleFactor).astype(np.uint8)

# Show disparity map
plt.imshow(dispImg,cmap="gray")
plt.show(block=False)
plt.pause(0.01)

# Save disparity map
cv.imwrite("disparity3_SGM.png",dispImg)

plt.show()
